File: playground-skywalking/order-service/src/main/java/com/currenjin/order/controller/OrderController.java
package com.currenjin.order.controller;

import com.currenjin.order.dto.OrderRequest;
import com.currenjin.order.dto.OrderResponse;
import com.currenjin.order.service.OrderService;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/orders")
public class OrderController {

    private final OrderService orderService;

    public OrderController(OrderService orderService) {
        this.orderService = orderService;
    }

    @PostMapping
    public OrderResponse placeOrder(@RequestBody OrderRequest request) {
        return orderService.placeOrder(request);
    }
}

File: poc-seata-at/src/main/kotlin/com/currenjin/seata/api/OrderController.kt
package com.currenjin.seata.api

import com.currenjin.seata.order.OrderService
import com.currenjin.seata.order.PlaceOrderCommand
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/orders")
class OrderController(
    private val orderService: OrderService,
) {
    @PostMapping
    fun create(@RequestBody request: CreateOrderRequest): ResponseEntity<CreateOrderResponse> {
        val orderId =
            orderService.placeOrder(
                PlaceOrderCommand(
                    userId = request.userId,
                    accountId = request.accountId,
                    amount = request.amount,
                    forceFail = request.forceFail,
                ),
            )

        return ResponseEntity.status(HttpStatus.CREATED).body(CreateOrderResponse(orderId))
    }
}

data class CreateOrderRequest(
    val userId: Long,
    val accountId: Long,
    val amount: Long,
    val forceFail: Boolean = false,
)

data class CreateOrderResponse(
    val orderId: Long,
)
